specs: ERC777 Stream 0.1.0 (#688)
RequestNetwork/requestNetwork: packages/advanced-logic/specs/payment-network-erc777-stream-0.1.0.md

> # Payment Network - ERC777 Stream

Reviewer: what is the name of the PN? `ERC777-stream` or `ERC777-fee-superfluid`

File name + title do not match the rest of the file

Reviewer 2: I've started with ERC777-fee-superfluid by adapting from ERC20-fee-proxy but ERC777-stream seems easier and more generic. Now that fees are by default it seems redundant to add it to the name. And for those who don't know yet SuperFluid stream is more expressive thand superfluid.

Author: Updated everything to `ERC777-stream` 👍

## Description

This extension allows payments to be made with streams of ERC777 Super tokens (cf. Superfluid Finance).
This payment network inherits from the [ERC20 Fee Proxy Contract](./payment-network-erc20-proxy-contract-0.1.0.md) extension.

> Many requests can be paid with the same stream, typically recurring requests of fix amount paid continuously. A group of requests payable with the same stream are called a request series, they must all have the same currency. It is possible to pay one series of requests with many streams, but it is not the purpose.

Reviewer: ```suggestion
Many requests can be paid with the same stream, typically recurring requests of fixed amount paid continuously. A group of requests payable with the same stream are called a request series, they must all have the same currency. It is possible to pay one series of requests with many streams, but it is not the purpose.
```

There are 2 kinds of request extensions in order to accept ERC777:

- The first request of a series is very similar to [payment-network-erc20-fee-proxy](./payment-network-erc20-fee-proxy-contract-0.1.0.md), it defines the salt, paymentAddress and requestId to compute the `paymentReference` used for the whole series.
- Other requests must define a `previousRequestId` and cannot redefine any of the payment properties (`paymentAddress`, `feeAddress` or `salt`).

> This version 0.1.0 does not allow refunds to be streamed. Refunds have to be made according to [payment-network-erc20-fee-proxy](./payment-network-erc20-fee-proxy-contract-0.1.0.md), using the fact that ERC777 tokens implement the ERC20 interface.

Reviewer 2: Cool, that should help to make this first version easier to do.

> ?? remaining questions ??

Reviewer: reminder

- impact on reputation
- flow rate < invoicing rate ?

As for [ERC20 Fee Proxy Contract](./payment-network-erc20-proxy-contract-0.1.0.md) requests, the request issuer can also declare payments manually. Fees shall not be paid for declarative payments.

> The SuperApp contract does the ERC777 token stream on behalf of the user. The contract ensures a link between an ERC777 SuperFluid stream and a request through a `paymentReference` stored in the userdata of the incoming stream. This `paymentReference` consists of the last 8 bytes of a salted hash of the requestId: `last8Bytes(hash(lowercase(requestId + salt + address)))`:

Reviewer: "does" is a bit vague for specs. Maybe process? or whatever is relevant for a token stream

As a payment network, this extension describes how to deduce a payment `balance` for the request. (see [Interpretation](#Interpretation))

## Payment SuperApp Contract

The contract contains one function called `streamFromWithReferenceAndFee` which takes 6 arguments:

- `tokenAddress` is the address of the ERC777 contract
- `to` is the destination address for the tokens
- `amount` is the amount of tokens to stream to the destination address
- `paymentReference` is the reference data used to track the stream (see `paymentReference`)
- `feeAmount` is the amount of tokens to stream to the fee destination address
- `feeAddress` is the destination address for the fee

The `StreamWithReferenceAndFee` event is emitted when the tokens are streamed. This event contains the same 6 arguments as the `streamFromWithReferenceAndFee` function.

> [See smart contract source](https://github.com/RequestNetwork/requestNetwork/blob/master/packages/smart-contracts/src/contracts/ERC777FeeSuperApp.sol)

Reviewer: 404

Reviewer 2: For the first version we don't use yet a SuperApp that's why this smart contract doesn't exist yet, We use first the SuperHost contract of SuperFluid.

| Network | Superhost Address                          |
| ------- | ------------------------------------------ |
| Private | FIXME                                      |
| matic   | 0x1EB3FAA360bF1f093F5A18d21f21f13D769d044A |
| xDai    | 0x8e8F05f1aD16D20e66Bd0922b510332104ddAc7B |

## Manual payment declaration

Manual payment declarations follow the same specifications as [payment-network-erc20-fee-proxy](./payment-network-erc20-fee-proxy-contract-0.1.0.md).

> ## Properties

Author: CR hint: this is quite original, challenges welcome

Reviewer 2: it seems the easiest way to connect series of requests.

## Action: Creation of children requests in a series

### Parameters

|                                  | Type   | Description                                          | Requirement   |
| -------------------------------- | ------ | ---------------------------------------------------- | ------------- |
| **id**                           | String | Constant value: "pn-erc777-fee-superfluid"           | **Mandatory** |
| **type**                         | String | Constant value: "paymentNetwork"                     | **Mandatory** |
| **version**                      | String | Constant value: "0.1.0"                              | **Mandatory** |
| **parameters**                   | Object |                                                      |               |
| **parameters.previousRequestid** | String | RequestId of the previous request in the same series | **Mandatory** |

### Conditions

This action is valid if:

- The `previousRequestid` is not empty and references an existing request with the same `currency`
- No request created before references the same request (there cannot be siblings)

### Results

The extension state is created with the following properties:

> |  Property                    |  Value                                               |

Reviewer: this table is not properly formatted

> | ---------------------------- | ---------------------------------------------------- | ------------- |

Reviewer: (not sure)
```suggestion
| ---------------------------- | ---------------------------------------------------- | 
```

| **events**                   | Array with one 'create' event (see below)            |

The 'create' event in the extension state **events**:

|  Property      |  Value                |
| -------------- | --------------------- |
| **name**       | 'create'              |
| **parameters** | the parameters object |

---

## Action: addPaymentAddress

### Parameters

|                               | Type   | Description                                | Requirement   |
| ----------------------------- | ------ | ------------------------------------------ | ------------- |
| **id**                        | String | Constant value: "pn-erc777-fee-superfluid" | **Mandatory** |
| **action**                    | String | Constant value: "addPaymentAddress"        | **Mandatory** |
| **parameters**                | Object |                                            |               |
| **parameters.paymentAddress** | String | Blockchain address for the payment         | **Mandatory** |

### Conditions

This action is valid, if:

- The extension state with the id "pn-erc777-fee-superfluid" exists
- The signer is the `payee`

> - The extension value `previousRequestId` does not exist

Author: CR hint: these conditions are new compared to other PNs

### Conditions

This action is valid, if:

- The extension state with the id "pn-erc777-fee-superfluid" exists
- The signer is the `payee`
- The extension value `previousRequestId` does not exist
- The extension value `feeAddress` is undefined
- The extension value `feeAmount` is undefined or represents an integer greater or equal than zero

### warnings

None.

### Results

The extension state is updated with the following properties:

|  Property             |  Value                                   |
| --------------------- | ---------------------------------------- |
| **values.feeAddress** | `feeAddress` from parameters             |
| **values.feeAmount**  | `feeAmount` from parameters              |
| **events**            | Add a 'fee' event (see below) at its end |

the 'addFee' event:

|  Property                 |  Value                          |
| ------------------------- | ------------------------------- |
| **name**                  | Constant value: "addFeeAddress" |
| **parameters**            |                                 |
| **parameters.feeAddress** | `feeAddress` from parameters    |
| **parameters.feeAmount**  | `feeAmount` from parameters     |

> ## Action: declareReceivedPayment TODO : assign to request or to suite?

Author: TODO section, in progress while you give me feedback for the main part

| **parameters.amount**  | Amount | The amount declared as received, in request currency          | **Mandatory** |
| **parameters.note**    | String | Additional information about the payment                      | Optional      |
| **parameters.txHash**  | String | The transaction hash for documentation and metadata           | Optional      |
| **parameters.network** | String | The network of the transaction for documentation and metadata | Optional      |

### Conditions

This action is valid, if:

- The extension state with the id "pn-erc777-fee-superfluid" exists
- The signer is the `payee`

### warnings

None.

### Results

An event is added to the extension state events array:

|  Property             |  Value                                   |
| --------------------- | ---------------------------------------- |
| **name**              | Constant value: "declareReceivedPayment" |
| **parameters**        |                                          |
| **parameters.amount** | `amount` from parameters                 |
| **parameters.note**   | `note` from parameters                   |
| **parameters.txHash** | `txHash` from parameters or undefined    |

> ## Action: declareReceivedRefund TODO : assign to request or to suite?

Author: TODO section, in progress while you give me feedback for the main part

| **parameters.note**    | String | Additional information about the payment                      | Optional      |
| **parameters.txHash**  | String | The transaction hash for documentation and metadata           | Optional      |
| **parameters.network** | String | The network of the transaction for documentation and metadata | Optional      |

### Conditions

This action is valid, if:

- The extension state with the id "pn-erc777-fee-superfluid" exists
- The signer is the `payee`

### warnings

None.

### Results

An event is added to the extension state events array:

|  Property              |  Value                                  |
| ---------------------- | --------------------------------------- |
| **name**               | Constant value: "declareReceivedRefund" |
| **parameters**         |                                         |
| **parameters.amount**  | `amount` from parameters                |
| **parameters.note**    | `note` from parameters                  |
| **parameters.txHash**  | `txHash` from parameters or undefined   |
| **parameters.network** | `network` from parameters or undefined  |

---

> ## Interpretation

Author: CR hint: brace yourself, this is the main part where I hope you can challenge the approach

Reviewer 2: Most of the times just checking the previous request in the series should be enough. I will challenge that.

### Fluctuating balance

The balance of an ongoing payment flow changes every second. Hence, to compute a balance that keeps in sync, the advanced-logic returns:

- `lastUpdatedBalance`
- `lastUpdateTimestamp`
- `currentFlowRate`

> We can only use the suite balance as a fixed number if the flow rate is zero.

Reviewer: I don't understand what this means

Author: OK, I will precise. This is to highlight that the balance can be a fixed number or a function of time, because it keeps changing.

Author: I clarified by describing the balance as a function of time !

The balance of one request cannot be greater than its expected amount except if we know that there is no more subsequent request in the series. Said another way: only the last request of the series can be overpaid.

In edge case scenarios, a new request is added to the end of the suite after an overpayment. The formerly last request becomes second to last, its balance is reduced to its `expectedAmount` and the remaining balance is allocated to the new request. Said another way: we can pay a suite before it is fully created, and the overpayment status of a suite can be temporary.

### Events

Three types of events have to be interpreted for the suite balance update:

- `FlowUpdated` when payment flows are created, updated or stopped, with agreements "Instant Distribution Agreement" or "Constant Flow Agreement"

> - ERC20 payments and refunds: inheriting from (payment-network-erc20-fee-proxy)[(./payment-network-erc20-fee-proxy-contract-0.1.0.md)]

Reviewer: ```suggestion
- ERC20 payments and refunds: inheriting from [payment-network-erc20-fee-proxy](./payment-network-erc20-fee-proxy-contract-0.1.0.md)
```

- Manual payment and refund declarations

> The SuperApp contract address is determined by the `request.currency.network` (see (table)[#Contract] with SuperApp contract addresses.

Reviewer: ```suggestion
The SuperApp contract address is determined by the `request.currency.network` (see [table](#Contract) with SuperApp contract addresses).
```

## Properties

### First request of a series

| Property                     | Type      | Description                                      | Requirement   |
| ---------------------------- | --------- | ------------------------------------------------ | ------------- |
| **id**                       | String    | constant value: "pn-erc777-stream"               | **Mandatory** |
| **type**                     | String    | constant value: "paymentNetwork"                 | **Mandatory** |
| **version**                  | String    | constant value: "0.1.0"                          | **Mandatory** |
| **events**                   | Array     | List of the actions performed by the extension   | **Mandatory** |
| **values**                   | Object    |                                                  |               |
| **values.salt**              | String    | Salt for the request                             | **Mandatory** |
| **values.expectedFlowRate**  | Number    | Expected amount of request `currency` per second | Optional      |

> | **values.expectedStartDate** | Timestamp | Expected start of stream                         | Optional      |

Reviewer 2: ```suggestion
| **values.expectedEndDate** | Timestamp | Expected deadline for start of stream                         | Optional      |
```

| **type**                     | String | constant value: "paymentNetwork"                | **Mandatory** |
| **version**                  | String | constant value: "0.1.0"                         | **Mandatory** |
| **events**                   | Array  | List of the actions performed by the extension  | **Mandatory** |
| **values**                   | Object |                                                 |               |
| **values.previousRequestId** | String | requestId of the previous request in the series | **Mandatory** |

---

## Action: Creation of the first request of a series

### Parameters

|                                 | Type & Description | Requirement   |
| ------------------------------- | ------------------ | ------------- |
| **id**                          | cf. Properties     | **Mandatory** |
| **type**                        | \*                 | **Mandatory** |
| **version**                     | \*                 | **Mandatory** |
| **parameters**                  | Object             |               |
| **parameters.salt**             | \*                 | **Mandatory** |

> | **parameters.expectedFlowRate** | \*                 | Optional      |

Author: Missing expectedStartDate
